Add interface altname support

kaffarell opened the pull request:
The linux kernel added interface altnames a few years back: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7a56493f0620.

Now each interface can have multiple altnames, and all these names, including the primary ones, must be unique. While the normal interface names are limited to 16 bytes, altnames can be up to 128 bytes long.

This update allows FRR to configure interfaces using these altnames. Zebra uses a red-black tree (RB tree) to store interfaces, which allows for quick lookups using the primary interface name. However, with the addition of altnames, this method isn't sufficient because we now have extra names for each interface.

To handle this, the system first checks for the primary name by quickly traversing the RB tree. If no primary name matches, it then goes through all interfaces one by one to check for matching altnames. This way, we maintain efficiency while support altnames.

Fixes: #6786

kaffarell committed:
zebra: consider altnames on iface lookup when compiled with flag

If the `--enable-transparent-altnames` option is passed when configuring
frr also consider altnames when the interface lookup functions are used.

Altnames are stored using a red-black tree on each interface. This means
that if we want to get an interface by its name we need to first check
the ifaces_by_name red-black tree and traverse it. If no interface is
found, we need to iterate linearly through all interfaces and check the
interface red-black tree of all the altnames.

This will decrease performance if the interface isn't found or if the
name is a altname. (Because in that case we need to iterate through all
interfaces linearly.)

Signed-off-by: Gabriel Goller <g.goller@proxmox.com>

kaffarell committed:
tests: add topotest for zebra altnames

Adds two altnames to an interface and checks if zebra recognizes and
shows them. Then add ip addrs to the same interface using the real name
and an altname.

Signed-off-by: Gabriel Goller <g.goller@proxmox.com>

kaffarell committed:
zebra: add altname support

Retrieve altnames from netlink and store them in the interface struct.
Show altnames when using `show interface` and `show interface json`
commands.

Get altnames from netlink, then store in them in the dplane_intf_info
(Interface info from the dataplane) struct in a RB-tree. Then copy them
over to the main interface struct where they are again stored in a
RB-tree.

Signed-off-by: Gabriel Goller <g.goller@proxmox.com>

ton31337 commented on lib/if.c:
```diff
@@ -474,6 +477,25 @@ ifindex_t ifname2ifindex(const char *name, vrf_id_t vrf_id)
+		/* Check altnames */
+		if (ifp->altnames) {
+			for (ALL_LIST_ELEMENTS_RO(ifp->altnames, node, altname)) {
+				if (strcmp(altname, name) == 0) {
```
nit: strmatch()

kaffarell replied:
obsolete.

ton31337 commented on zebra/if_netlink.c:
```diff
@@ -1355,6 +1361,33 @@ int netlink_link_change(struct nlmsghdr *h, ns_id_t ns_id, int startup)
+				if (alt_attr->rta_type == IFLA_ALT_IFNAME) {
+					char *altname = XSTRDUP(MTYPE_TMP, RTA_DATA(alt_attr));
+
+					if (altname)
```
How this can be NULL?

kaffarell replied:
it can't, my bad.

ton31337 commented on zebra/if_netlink.c:
```diff
@@ -66,6 +66,12 @@
 extern struct zebra_privs_t zserv_privs;
 
+/* Helper function to free altname strings */
+static void altname_free(void *data)
```
Why is this method in this file (netlink)?

kaffarell replied:
oops, missed this, isn't used anymore.

ton31337 commented on zebra/interface.c:
```diff
@@ -3059,6 +3105,18 @@ static void if_dump_vty_json(struct vty *vty, struct interface *ifp,
+		for (ALL_LIST_ELEMENTS_RO(ifp->altnames, inner_node, altname))
+			json_object_array_add(json_altnames, json_object_new_string(altname));
+
+		json_object_object_add(json_if, "altnames", json_altnames);
```
We should use camelCase format: `altNames`.

ton31337 reviewed (commented):
One question, though: why can't we use a tree and use a list instead?

mjstapp reviewed (commented):
my top-level question is about what the use-case is for supporting these names. I don't have an objection to capturing the OS's info and making it visible, if that's useful to someone. but I'm concerned about introducing ambiguity if we go further than that.

ton31337 commented on zebra/zebra_dplane.h:
```diff
@@ -293,6 +294,9 @@ PREDECL_DLIST(dplane_ctx_list);
 PREDECL_DLIST(dplane_intf_extra_list);
 
+RB_HEAD(altnames_dplane_head, altname);
+RB_PROTOTYPE(altnames_dplane_head, altname, entry, altname_cmp_func);
```
@mjstapp previous version was a list and I suggested doing the same as for interface (so I blame myself here) :)

mjstapp replied:
heh - got it!
I think the top-level question is whether we're going to treat these alt names as if they were "real" names. if not, if we're adding just a list of strings, then maybe we could use a single definition for that list in both the interface and dplane objects.

kaffarell commented on lib/if.c:
```diff
@@ -389,6 +396,15 @@ void if_delete(struct interface **ifp)
 	XFREE(MTYPE_IFDESC, ptr->desc);
 
+	/* Clean up altnames list */
+	struct altname *item;
```
Thanks for the review.

mjstapp commented on lib/if.h:
```diff
@@ -206,6 +207,21 @@ struct if_link_params {
+/* Length of interface altname.*/
+#define IF_ALTNAMESIZE 128
+
+struct altname {
```
but ... I think from the management system pov, this list of aliases is "state" - it's an attribute that arrives at frr from the OS. I see that your own test doesn't try to do the thing you just described there - configure the same object as "interface x" and "interface y".  The yang model for interfaces uses a single string key to identify an interface config object. If you think that aliased configuration path is worthwhile, I think you'll have to figure out how to model that. I'm not convinced it's worthwhile - but I'd be interested to hear from other folks.

kaffarell replied:
> I see that your own test doesn't try to do the thing you just described there - configure the same object as "interface x" and "interface y"

True, I'll add that as a testcase.

>  The yang model for interfaces uses a single string key to identify an interface config object.

I've never used yang/netconf, so please bear with me. I skimmed over rfc8343 and to me it seems like this should be allowed? Every altname is still a unique identifier for a specific interface. @ton31337 what do you think?

kaffarell commented:
Added a `interface x` and `interface <altname of x>` topotest. Also removed some character length limitations that prevented `interface altnamelongerthan15chars` to work with yang.

kaffarell committed:
docs: zebra: add mention of --enable-transparent-altnames flag

Add short mention of the --enable-transparent-altnames flag which
changes the way `interface IFNAME` looks up names (it includes
altnames).

Signed-off-by: Gabriel Goller <g.goller@proxmox.com>

kaffarell commented:
Just pushed a new version. As discussed in last weeks tech meeting, I added a configuration flag `--enable-transparent-altnames` (no hard feelings on the name) which enables or disables the lookup of altnames. I did not put the altname parsing and displaying (in `show interface`) behind the configuration flag as I think thats not really an issue.

kaffarell commented on lib/if.h:
```diff
@@ -206,6 +207,21 @@ struct if_link_params {
+/* Length of interface altname.*/
+#define IF_ALTNAMESIZE 128
+
+struct altname {
```
Pushed a new version adding yang support (first commit). Thanks.

riw777 commented:
What did we decide on this one?

riw777 reviewed (dismissed):
looks good ... waiting on comments from @mjstapp ??

kaffarell commented:
Rebased this PR.
